File: TensorRT/python/trt_run.py
import os
import numpy as np
import tensorrt as trt

from .utils import common, calibrator


class TRTModel:
    def __init__(self, onnx_path, plan_path, mode="fp16"):
        """
        :param onnx_path: local path of onnx file.
        :param plan_path: trt plan file to read/save.
        :param mode: inference mode, fp16/int8.
        """
        self.trt_logger = trt.Logger()

        self.onnx_path = onnx_path
        self.plan_path = plan_path
        self.mode = mode

        self.parse_input_shape = []
        self.parse_output_shape= []

        self.engine = self._get_engine()
        self.execution_context = self.engine.create_execution_context()
        self.inputs, self.outputs, self.bindings, self.stream = common.allocate_buffers(self.engine)

    def _check_network(self, network):
        """check network
        :param network: INetworkDefinition
        """
        if not network.num_outputs:
            raise Exception("No output node found!")

        input_nodes = [network.get_input(i) for i in range(network.num_inputs)]
        output_nodes = [network.get_output(i) for i in range(network.num_outputs)]

        print("Network description")
        for i, inp in enumerate(input_nodes):
            print("Input node {} | Name {} | Shape {}".format(i, inp.name, inp.shape))

        print("Total layers: {}".format(network.num_layers))
        for i in range(network.num_layers):
            layer = network.get_layer(i)
            print("index {}, layer name: {}".format(i, layer.name))

        for i, out in enumerate(output_nodes):
            print("Output node {} | Name {} | Shape {}".format(i, out.name, out.shape))

    def _parse_onnx(self):
        """takes an ONNX file and creates a TensorRT engine to run inference with
        """
        dynamic = False
        flag = common.EXPLICIT_BATCH

        with trt.Builder(self.trt_logger) as builder, builder.create_network(flag) as network, builder.create_builder_config() as config, trt.OnnxParser(network, self.trt_logger) as parser, trt.Runtime(self.trt_logger) as runtime:
            config.max_workspace_size = common.GiB(1)
            builder.max_batch_size = 1

            if self.mode == "fp16":
                config.set_flag(trt.BuilderFlag.FP16)
                print("set FP16 mode.")
            if self.mode == "int8":
                config.set_flag(trt.BuilderFlag.INT8)
                print("set INT8 mode.")
            
        # Parse model file
        print('Loading ONNX file from path {}...'.format(self.onnx_path))
        with open(self.onnx_path, 'rb') as model:
            print('Beginning ONNX file parsing')
            if not parser.parse(model.read()):
                print('ERROR: Failed to parse the ONNX file.')
                for error in range(parser.num_errors):
                    print(parser.get_error(error))
                return None
        print('Completed parsing of ONNX file')

        # check netowrk
        self._check_network(network)

        # build engine
        print('Building an engine from file {}; this may take a while...'.format(self.onnx_path))
        plan = builder.build_serialized_network(network, config)
        engine = runtime.deserialize_cuda_engine(plan)
        print("Completed creating Engine")

        # save engine
        with open(self.plan_path, "wb") as f:
            f.write(plan)

        return engine

    def _get_engine(self):
        """generate tensorrt runtime engine
        """
        if os.path.exists(self.plan_path):
            print('Load trt plan from: {}'.format(self.plan_path))

            with open(self.plan_path, "rb") as f, trt.Runtime(self.trt_logger) as runtime:
                return runtime.deserialize_cuda_engine(f.read())
        else:
            if os.path.exists(self.onnx_path):
                return self._parse_onnx()
            else:
                raise Exception("ONNX model file {} not exist!".format(self.onnx_path))

    def forward(self, image_tensors):
        """do infernece
        :param onnx_path: list, inputs tensor of model.
        :return outputs: list, outputs tensor of model.
        """

        for i, image_tensor in enumerate(image_tensors):
            image = np.array([image_tensor], dtype=np.float32, order='C')
            self.inputs[i].host = image

        trt_outputs = common.do_inference_v2(self.execution_context, 
                                             bindings=self.bindings,
                                             inputs=self.inputs,
                                             outputs=self.outputs,
                                             stream=self.stream)

        return trt_outputs

File: TensorRT/cpp/src/trt_run.cpp
#include "trt_run.h"


/**
 * @brief Construct a new TRTModel::TRTModel object
 * 
 * @param devId id of cuda device to run.
 * @param modelPath local path to *.onnx model file
 * @param planPath local path to *.plan model file, read/save.
 * @param mode inference mode, fp16/int8
 * @param batchSize max batch_size of model
 * @param useDLACore if use nvidia dla core
 */
TRTModel::TRTModel(int devId, const string& modelPath, const string& planPath, const string &mode, int batchSize, const string& calDataFileName, const string& calTableName, int useDLACoreIndex):
        modelPath(modelPath), planPath(planPath), mode(mode), batchSize(batchSize), useDLACoreIndex(useDLACoreIndex), calDataFileName(calDataFileName), calTableName(calTableName),
        mEngine(nullptr), mStream(nullptr), mContext(nullptr) {
    cudaSetDevice(devId);
    cout << "set cuda device id: " << to_string(devId) << endl;

    bool status = false;

    if (fileExistCheck(this->planPath)) {
        cout << "load trt engine from plan file: " << this->planPath << endl;
        this->loadEngine();
    } else {
        status = this->constructModel();
        assert(status);
    }

    status = this->init();
    assert(status);
}

/**
 * @brief create trt engine from onnx.
 * 
 */
bool TRTModel::constructModel() {
    // create engine builder
    auto builder = SampleUniquePtr<nvinfer1::IBuilder>(nvinfer1::createInferBuilder(this->trtLogger));
    if (!builder)
        return false;
    
    // create engine network
    const auto explicitBatch = 1U << static_cast<uint32_t>(nvinfer1::NetworkDefinitionCreationFlag::kEXPLICIT_BATCH);
    auto network = SampleUniquePtr<nvinfer1::INetworkDefinition>(builder->createNetworkV2(explicitBatch));
    if (!network)
        return false;
    
    // create ONNX parser 
    auto parser = SampleUniquePtr<nvonnxparser::IParser>(nvonnxparser::createParser(*network, this->trtLogger));
    if (!parser)
        return false;
    
    // parse Onnx model and save to network
    auto parsed = parser->parseFromFile(this->modelPath.c_str(), static_cast<int>(nvinfer1::ILogger::Severity::kWARNING));
    if (!parsed)
        return false;
    
    this->checkNetwork(network.get());

    // get engine config from builder
    auto config = SampleUniquePtr<nvinfer1::IBuilderConfig>(builder->createBuilderConfig());
    if (!config)
        return false;
    
    // config
    builder->setMaxBatchSize(this->batchSize);
    config->setMaxWorkspaceSize(1 << 30);

    // Calibrator life time needs to last until after the engine is built.
    unique_ptr<nvinfer1::IInt8Calibrator> calibrator;

    if (this->mode == "fp16") {
        bool isSupport = builder->platformHasFastFp16();
        if (!isSupport)
            cout << "this device may not fast on fp16 mode." << endl;
        config->setFlag(nvinfer1::BuilderFlag::kFP16);
        cout << "set fp16" << endl;
    }

    if (this->mode == "int8") {
        bool isSupport = builder->platformHasFastInt8();
        if (!isSupport) {
            cout << "this device may not fast on int8 mode." << endl;
            config->setFlag(nvinfer1::BuilderFlag::kFP16);
            cout << "set fp16" << endl;
        } else {
            config->setFlag(nvinfer1::BuilderFlag::kINT8);

            // set int8 calibrator
            calibrator.reset(new Int8EntropyCalibrator(this->calDataFileName, this->calTableName, this->batchSize, this->inputWidth, this->inputHeight));
            config->setInt8Calibrator(calibrator.get());
            cout << "set int8" << endl;
        }
    }

    // set DLA if use jetson device
    if (this->useDLACoreIndex != -1) {
        cout << "use DLA core index: " << to_string(this->useDLACoreIndex) << endl;
        this->enableDLA(builder.get(), config.get());
    }

    // build trt engine
    SampleUniquePtr<nvinfer1::IHostMemory> plan{builder->buildSerializedNetwork(*network, *config)};
    if (!plan)
        return false;
    
    SampleUniquePtr<nvinfer1::IRuntime> runtime{nvinfer1::createInferRuntime(this->trtLogger)};
    if (!runtime)
        return false;

    this->mEngine = std::shared_ptr<nvinfer1::ICudaEngine>(
        runtime->deserializeCudaEngine(plan->data(), plan->size()), InferDeleter());
    if (!this->mEngine)
        return false;

    this->saveEngine();

    return true;
}

/**
 * @brief init trt runtime.
 * 
 */
bool TRTModel::init() {
    // create cuda stream
    CHECK(cudaStreamCreate(&this->mStream));
    // create cuda execute context
    this->mContext = shared_ptr<nvinfer1::IExecutionContext>(this->mEngine->createExecutionContext(), InferDeleter());

    // allocate cuda memory of input node and output node
    for (int i = 0; i < this->mEngine->getNbBindings(); i++) {
        nvinfer1::Dims dims = this->mEngine->getBindingDimensions(i);
        nvinfer1::DataType dtype = this->mEngine->getBindingDataType(i);
        string bindName = this->mEngine->getBindingName(i);

        int64_t elemNum = volume(dims);
        int64_t nbBytes = elemNum * getElementSize(dtype);

        auto *buffer = new CudaBuffer(i, nbBytes, elemNum);
        if (this->mEngine->bindingIsInput(i)) {
            this->inputBind.push_back(buffer);
        } else {
            this->outputBind.push_back(buffer);
        }
    }

    return true;
}

/**
 * @brief check network structure.
 * 
 * @param network trt network.
 */
void TRTModel::checkNetwork(nvinfer1::INetworkDefinition *network) {
    cout << "=== Network Description ===" << endl;

    for (int i = 0; i < network->getNbInputs(); i++) {
        auto input = network->getInput(i);
        auto dims = input->getDimensions();
        this->inputWidth = dims.d[2];
        this->inputHeight = dims.d[1];
        string shape = "(" + to_string(dims.d[0]) + ", " + to_string(dims.d[1]) + ", " + to_string(dims.d[2])  + ", " + to_string(dims.d[3]) + ")" ;
        cout << "Input: " << to_string(i) << " | Name: " << input->getName() << " | shape: " << shape << endl;
    }

    for (int i = 0; i < network->getNbOutputs(); i++) {
        auto output = network->getOutput(i);
        auto dims = output->getDimensions();
        string shape = "(" + to_string(dims.d[0]) + ", " + to_string(dims.d[1]) + ", " + to_string(dims.d[2]) + ")" ;
        cout << "Output: " << to_string(i) << " | Name: " << output->getName() << " | shape: " << shape << endl;
    }
}

/**
 * @brief set dla in trt engine.
 * 
 * @param builder trt builder.
 * @param config trt config.
 * @param allowGPUFallback if allow GPU Fallback
 */
void TRTModel::enableDLA(nvinfer1::IBuilder* builder, nvinfer1::IBuilderConfig* config, bool allowGPUFallback) {
    int coreNum = builder->getNbDLACores();

    if (coreNum == 0) {
        cerr << "Trying to use DLA core on a platform that doesn't have any DLA cores" << endl;
        assert("Error: use DLA core on a platfrom that doesn't have any DLA cores" && false);
    }

    if (allowGPUFallback) {
        config->setFlag(nvinfer1::BuilderFlag::kGPU_FALLBACK);
    }

    if (!config->getFlag(nvinfer1::BuilderFlag::kINT8)) {
        // User has not requested INT8 Mode.
        // By default run in FP16 mode. FP32 mode is not permitted.
        config->setFlag(nvinfer1::BuilderFlag::kFP16);
    }

    config->setDefaultDeviceType(nvinfer1::DeviceType::kDLA);
    config->setDLACore(this->useDLACoreIndex);
    config->setFlag(nvinfer1::BuilderFlag::kSTRICT_TYPES);
}

/**
 * @brief load engine from plan file.
 * 
 */
bool TRTModel::loadEngine() {
    bool status = false;

    if (!this->planPath.empty()) {
        ifstream fin(this->planPath);

        // 将文件中的内容读取至字符串
        string cachedEngine;
        while (fin.peek() != EOF) { // 使用fin.peek()防止文件读取时无限循环
            stringstream buffer;
            buffer << fin.rdbuf();
            cachedEngine.append(buffer.str());
        }
        fin.close();

        SampleUniquePtr<nvinfer1::IRuntime> runtime{nvinfer1::createInferRuntime(this->trtLogger)};
        this->mEngine = std::shared_ptr<nvinfer1::ICudaEngine>(
            runtime->deserializeCudaEngine(cachedEngine.data(), cachedEngine.size()), InferDeleter());

        if (!this->mEngine)
            status = true;
    }

    return true;
}

/**
 * @brief save engine to plan file.
 * 
 */
bool TRTModel::saveEngine() {
    bool status = false;

    if (!this->planPath.empty()) {
        auto data = SampleUniquePtr<nvinfer1::IHostMemory>(this->mEngine->serialize());

        ofstream serializeOutputStream;

        // 将序列化的模型结果拷贝至serialize_str字符串
        string serializeStr;
        serializeStr.resize(data->size());
        memcpy((void*)serializeStr.data(), data->data(), data->size());

        // 将serialize_str字符串的内容输出至文件
        serializeOutputStream.open(this->planPath);
        serializeOutputStream << serializeStr;
        serializeOutputStream.close();

        status = true;
    }

    return status;
}

/**
 * @brief copy data between host and cuda.
 * 
 */
void TRTModel::copyData(void *dstPtr, const void *srcPtr, int64_t byteSize, cudaMemcpyKind memcpyType, bool async, const cudaStream_t &stream) {
    if (async)
        CHECK(cudaMemcpyAsync(dstPtr, srcPtr, byteSize, memcpyType, stream));
    else
        CHECK(cudaMemcpy(dstPtr, srcPtr, byteSize, memcpyType));
}

/**
 * @brief set dla in trt engine.
 * 
 * @param inputData input data.
 * @param outputData output data.
 */
void TRTModel::forward(const vector<float>& inputs, vector<float>& output) {
    // copy data from host to cuda
    vector<void*> inputDataList;
    for (const auto &ib: this->inputBind) {
        this->copyData(ib->getData(), inputs.data(), ib->getBytes(), cudaMemcpyHostToDevice, true, this->mStream);
        inputDataList.emplace_back(ib->getData());
    }
    for (const auto &ob: this->outputBind) {
        inputDataList.emplace_back(ob->getData());
    }

    // inference
    this->mContext->executeV2(inputDataList.data());

    // copy data from cudao to host
    output.resize(this->outputBind[0]->getElements());
    this->copyData(output.data(), this->outputBind[0]->getData(), this->outputBind[0]->getBytes(), cudaMemcpyDeviceToHost, true, this->mStream);
    
    // sync cuda stream
    cudaStreamSynchronize(this->mStream);
}

/**
 * @brief Destroy the TRTModel::TRTModel object
 * 
 */
TRTModel::~TRTModel() {
    cudaStreamSynchronize(this->mStream);
    cudaStreamDestroy(this->mStream);

    for (auto& item: this->inputBind)
        delete item;
    for (auto& item: this->outputBind)
        delete item;
}
